fix(core): prevent connection reset during inactivity

jerrinot opened the pull request:
This is best-effort.

puzpuzpuz commented:
> This is best-effort. For now implemented on Linux only.

Is the description still actual? I can see some flags being set on other OSes, not only Linux.

puzpuzpuz commented:
The BSD binary is missing.

puzpuzpuz commented on core/src/main/c/share/net.c:
```diff
@@ -43,17 +43,31 @@ int set_int_sockopt(int fd, int level, int opt, int value) {
-        }
+        // enable TCP keepalive probes, otherwise firewalls and load balancers may drop connections
+        // during a period of inactivity
+        set_int_sockopt(fd, SOL_SOCKET, SO_KEEPALIVE, 1);
```
nit: we should check errors here and below. Otherwise, we won't notice if these calls do nothing instead of the intended action.

jerrinot replied:
I was considering that. but what do you if the options cannot be set? fail questdb? I thought best-effort is OK in this case. wdyt?

puzpuzpuz replied:
Well, at least we could log something. But we could even fail the socket creation as a whole - that's to detect compatibility issues (very unlikely to happen).

jerrinot replied:
I was considering creating a separate private native function `Net.setKeepAlive()` and call it from `socketTcp()` and guard it with a system property. Then it would make sense to check return values - because if on some OS the options cannon be set then we could use the system property as a fallback and disable it. But when it's done unconditionally on all sockets then I don't dare to fail socket creation just because KeepAlive cannot be set.

puzpuzpuz replied:
TBH I really like your initial approach, i.e. a separate method. I don't think an extra JNI call overhead matters when it comes to socket creation. WDYT?

jerrinot replied:
I was not worried about JNI overhead, I thought it was unnecessarily complex. but I can do it like if you find it better.

puzpuzpuz replied:
To me, we would be able at least to log errors with a standalone method. I don't like the idea of swallowing errors in the native code.

jerrinot commented:
@puzpuzpuz 
>>    This is best-effort. For now implemented on Linux only.

> Is the description still actual? I can see some flags being set on other OSes, not only Linux.

Good one! It was originally implemented for Linux-only, later I added Windows and Mac impl. 
There is no FreeBSD version that's the reason I did not rebuild the FreeBSD binary.

puzpuzpuz commented:
> There is no FreeBSD version that's the reason I did not rebuild the FreeBSD binary.

The BSD version should be identical to the OS X branch. If so, it should be simple to add it.

jerrinot commented:
@puzpuzpuz do you know the macro name I should test for FreeBSD? or perhaps there is something like DARWIN - which is set on both MacOS and FreeBSD?

puzpuzpuz commented:
> @puzpuzpuz do you know the macro name I should test for FreeBSD? or perhaps there is something like DARWIN - which is set on both MacOS and FreeBSD?

I guess, you use `#else` for the `#ifdef __linux__` macro branch. Other than that, BSD is `__FreeBSD__`. This website is a nice helper: https://sourceforge.net/p/predef/wiki/OperatingSystems/

puzpuzpuz commented on core/src/main/java/io/questdb/network/Net.java:
```diff
@@ -279,9 +283,18 @@ public static long sockaddr(int ipv4address, int port) {
 
     public static int socketTcp(boolean blocking) {
-        return Files.bumpFileCount(socketTcp0(blocking));
+        return Files.bumpFileCount(setKeepAlive(socketTcp0(blocking), TCP_KEEPALIVE_SECONDS));
```
This call chain doesn't look nice. If `socketTcp0()` is successful and `setKeepAlive()` fails, we'll return a negative value and leak the fd.

Better to keep them separate and maybe log something if `setKeepAlive()` fails instead of failing the fd creation.

jerrinot replied:
> we'll return a negative value and leak the fd. 

that's a good point, thanks!

jerrinot replied:
how about now?

jerrinot commented:
fwiw: it looks FreeBSD requires different treatment than Apple:
```
/home/questbuild/local/java/questdb/core/src/main/c/share/net.c:57:46: error: use of undeclared identifier 'TCP_KEEPALIVE'
        if (set_int_sockopt(fd, IPPROTO_TCP, TCP_KEEPALIVE, idle_sec) < 0) {
                                             ^
1 error generated.
gmake[2]: *** [CMakeFiles/questdb.dir/build.make:212: CMakeFiles/questdb.dir/src/main/c/share/net.c.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:246: CMakeFiles/questdb.dir/all] Error 2
gmake: *** [Makefile:114: all] Error 2
```
:(

puzpuzpuz commented:
> fwiw: it looks FreeBSD requires different treatment than Apple:

Yeah, looks like the flag is SO_KEEPALIVE: https://man.freebsd.org/cgi/man.cgi?setsockopt

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 7 / 10 (70.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/network/Net.java|7|10|70.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 7 / 10 (70.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/network/Net.java|7|10|70.00%|

puzpuzpuz commented on core/src/main/c/share/net.c:
```diff
@@ -40,6 +40,27 @@ int set_int_sockopt(int fd, int level, int opt, int value) {
+    if (set_int_sockopt(fd, SOL_SOCKET, SO_KEEPALIVE, 1) < 0) {
+        return -1;
+    }
+    #if defined(__linux__) || defined(__FreeBSD__)
```
nit: could be `#if defined(__unix__)`

jerrinot replied:
but is it really better? is MacOS a Unix? Probably more than Linux.

puzpuzpuz replied:
> is MacOS a Unix?

Not from the macro perspective. :)

Anyway, let's keep it as is.

puzpuzpuz commented on core/src/main/java/io/questdb/network/Net.java:
```diff
@@ -279,9 +287,23 @@ public static long sockaddr(int ipv4address, int port) {
+        }
+        if (setKeepAlive0(fd, TCP_KEEPALIVE_SECONDS) < 0) {
+            int errno = Os.errno();
+            LOG.error().$("could not set tcp keepalive [fd=").$(fd).$(", errno=").$(errno).$(']').$();
```
nit:
```suggestion
            LOG.error().$("could not set tcp keepalive [fd=").$(fd).$(", errno=").$(errno).I$();
```

bluestreak01 commented:
@jerrinot what's the plan? Are we addressing comments or merging?

jerrinot commented:
@bluestreak01 I accepted one suggestion, but I am not sure about the other. so I'm for merging it, unless @puzpuzpuz insists on the change.

puzpuzpuz commented:
> I'm for merging it, unless @puzpuzpuz insists on the change.

"nit:" comments are disposable at will by the definition. That's why I approved the PR.

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 7 / 10 (70.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/network/Net.java|7|10|70.00%|

jerrinot commented:
@puzpuzpuz thanks for your review!

puzpuzpuz commented:
@jerrinot my pleasure!
